cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/BrokerFailureDetector.java

```diff
@@ -88,16 +88,22 @@ void startDetection() {
     _zkClient.subscribeChildChanges(ZkUtils.BrokerIdsPath(), new BrokerFailureListener());
   }
 
-  synchronized void detectBrokerFailures() {
+  private synchronized void detectBrokerFailures(Set<Integer> aliveBrokers) {
     // update the failed broker information based on the current state.
-    updateFailedBrokerList(aliveBrokers());
-    // persist the updated failed broker list.
-    persistFailedBrokerList();
+    boolean updated = updateFailedBrokers(aliveBrokers);
+    if (updated) {
+      // persist the updated failed broker list.
+      persistFailedBrokerList();
+    }
     // Report the failures to anomaly detector to handle.
     reportBrokerFailures();
   }
 
-  Map<Integer, Long> failedBrokers() {
+  synchronized void detectBrokerFailures() {
```

Comment:
Is this `synchronized` necessary?

Reply:
This ensures that if concurrent calls to `BrokerFailureDetector#detectBrokerFailures()` wait for `BrokerFailureDetector#detectBrokerFailures(Set<Integer>)`, then the call to `BrokerFailureDetector#aliveBrokers()` will be more fresh.